---
title: "100-generate-sample-list"
author: "Mac Campbell"
date: "July 5, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
```

## What do we have?

Dropping two known problematic samples
```{r}
metaall<-read_csv("metadata/genetics-oto-intersect-06032021_JH.csv") %>%
  filter(!(`Sequence File Name` %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10"))) %>%
  filter(aggregated_new_classes %in% c("FWR","MIG")) %>%
  filter(`Birth Year (HOBBS)`==2012)

write_csv(metaall, "metadata/2012-all-samples.csv")
metaall %>% group_by(aggregated_new_classes) %>% summarize(Count=n())
```

Let's round up those raw fastqs and align to the genome.     

__1__ Find samples     
__2__ Align    
__3__ Filter low-coverage samples    
__4__ Run association
For 1 & 2, see doAlign.sh for basics.

### Get number of reads/bam file.    

Number of reads, % mapped from .sort.bam files. Reverse enginerring from %mapped doesn't work well. 
/home/maccamp/ds-lh/data

`ls | grep sort.bam | grep -v bai | while read line; do samtools flagstat $line | grep total >> totals.txt; done;`
`ls | grep sort.bam | grep -v bai >> totals.files.txt`
`ls | grep sort.bam | grep -v bai | while read line; do samtools flagstat $line | grep "mapped (" >> mappeds.txt; done; `

` ls | grep sort.flt.bam | grep -v bai | while read line; do samtools flagstat $line | grep mapped | head -n 1 >> counts.txt; done;`     
` ls | grep sort.flt.bam | grep -v bai >> counts.files.txt`    

```{r}
files<-read_tsv("outputs/100/counts.files.txt", col_names="File")
counts<-read_tsv("outputs/100/counts.txt", col_names="Counts")
counts$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts$Counts, fixed = TRUE)

counted<-bind_cols(files,counts)
counted$`Sequence File Name`<-gsub("_R1.sort.flt.bam","",counted$File)
```

#### More info as requested by reviewer

```{r}
totals<-read_tsv("outputs/100/totals.files.txt", col_names=c("File"))
totalscounts<-read_tsv("outputs/100/totals.txt", col_names=c("Total"))
mapped<-read_tsv("outputs/100/mappeds.txt", col_names=c("Mapped"))
totals<-bind_cols(totals, totalscounts, mapped)

totals$`Sequence File Name`<-gsub("_R1.sort.bam","",totals$File)
totals<-totals %>% separate(Total, into=c("Total","Whatever1","Whatever2"), sep=" \\+") %>% select(-Whatever1, -Whatever2)
totals<-totals %>% separate(Mapped, into=c("Mapped","Whatever"), sep=" \\+") %>% select(-Whatever, -File)
totals$Mapped<-as.numeric(totals$Mapped)
totals$Total<-as.numeric(totals$Total) 

totals<- totals %>%
  mutate(`Percentage Mapped`= round((Mapped/Total)*100,2))
totals
```


```{r}
reads<-left_join(metaall, counted) %>%
  mutate(R1=paste0(`Sequence File Name`,"_R1")) %>% 
  mutate(R2=paste0(`Sequence File Name`,"_R2"))  %>%
  mutate(Pheno=ifelse(aggregated_new_classes=="MIG",0,1))
reads$Counts<-as.numeric(reads$Counts)

write_csv(reads, "outputs/100/reads.csv")
```

```{r}
#Create Supplemental Document S1
#joining by Ht20-33_2012_A05     `Sequence File Name`

sdoc<-reads %>% select(-X5,-X12,-X17, -X18, -X20) %>% rename(FishID=X1) %>% 
  rename(`Sex 1=male, 2=female, 0=unknown`=sex) %>% select(-old_classes, -change, -Pheno) %>%
  rename(`Aligned and deduplicated reads` = Counts) %>% left_join(totals)
write_csv(sdoc, "Supplemental/Supplemental Document S1.csv")
```

```{r}
tops<-reads %>% top_frac(.75, Counts) %>% filter(sex!=0)
samplesize<-tops %>% group_by(aggregated_new_classes ,sex) %>% summarize(`Sample Size`=n())
samplesize
```

```{r}
ggplot(tops) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Pheno) +
  xlim(0, max(tops$Counts))
```

```{r}
tops %>% group_by(aggregated_new_classes, sex) %>% select(Counts) %>% summarize_all(mean) %>% left_join(samplesize)
```

## Generate Bamlists, Phenotype and Sex Files

```{r}
bamlist<-tops %>% select(File) %>% mutate(Path=paste0("bams/",File)) %>% select(Path)
write_tsv(bamlist, "bamlists/2012.bamlist", col_names=FALSE)

phenos<-tops %>% select(Pheno) 
write_tsv(phenos, "phenos/2012.phenos", col_names=FALSE)

sex<-tops %>% select(sex)
write_tsv(sex, "phenos/2012.cov", col_names=FALSE)

write_csv(tops, "metadata/2012-samples.csv")
```

