---
title: "sanity-check"
author: "Mac Campbell"
date: "July 28, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)
```

```{r}
library(tidyverse)
library(adegenet)
library(snpR)
library(RColorBrewer)
library(ggrepel)
```

## How does DAPC work when we separate randomly
```{r}
load("outputs/figure-3/genind.rda")

gid<-genind

#Generate random assignments
vec <- round(runif(n = 121, min = 1, max = 2))
vec<-gsub(pattern = 1, replacement = "One", vec)
vec<-gsub(pattern = 2, replacement = "Two", vec)
vec<-as.factor(vec)
x<-vec

#x<-as.factor(rep(c("One","Two"),times=c(60,61)))

gid@pop<-x
```

```{r}
dapc<-dapc(gid, n.pca=110, n.da=1)

scatter(dapc, col=c("red","blue"))
pdf("outputs/figure-3/sanity-check.pdf", width = 6, height=4)

scatter(dapc, col=c("red","blue"))
dev.off()
```

```{r}
post<-as_tibble(dapc$posterior)
post$Phenotype<-dapc$grp
post <- post %>% mutate(FID=1:n()) %>% select(FID, Phenotype, One, Two)
tidy<-post %>% pivot_longer(!c(FID,Phenotype)) %>% rename(Assignment=name)

ggplot(tidy %>% filter(Phenotype==Assignment), aes(x=value, color=Assignment, fill=Assignment), alpha=0.9) +
  geom_histogram() +
  facet_wrap(.~Phenotype, ncol=1) +
  scale_color_manual(values=c("red","blue")) +
  scale_fill_manual(values=c("red","blue")) +
  theme_bw() +
  theme(panel.grid=element_blank())+
  theme(panel.background = element_blank()) +
  ylab("Count\n") +
  xlab("\nPosterior Probability") +
  theme(legend.position = "")

ggsave("outputs/figure-3/sanity-assign.pdf", width=6, height = 4)
```

```{r}
contrib <- loadingplot(dapc$var.contr, axis=1,
thres=.0005, lab.jitter=0)
```

## Do Asso
```{r}
genos<-read_tsv("outputs/100/snps-characters.geno.gz", col_names=FALSE)
genos<-genos[1:ncol(genos)-1]

genos$X1<-as.factor(genos$X1)

# grab our sample metadata
snp_meta <- genos[,1:2]
sample_meta<-x %>% as_tibble() %>% rename(Class=value)
sample_meta$Phenotype<-as.factor(sample_meta$Class)
# import, remember to remove metadata from the genotypes!
my.dat <- import.snpR.data(genos[,-c(1:2)], 
                           snp.meta = snp_meta, 
                           sample.meta = sample_meta)


asso<-calc_association(my.dat,response="Phenotype", method="chisq")
out<-get.snpR.stats(asso) %>% as_tibble()
```

```{r}
dd<-out %>% mutate(Index=1:n()) %>% rename(Chromosome=X1, Position=X2) %>% 
  mutate(bonfer = p.adjust(chi_p_Phenotype, method = "bonferroni")) 
dd$Chromosome<-as.factor(dd$Chromosome)

#calc significance
variants<-nrow(dd)
#Number of expected variants
num<-10
p.T = num/variants 
prior.odds = p.T/(1-p.T) 
pwr = 1 #upper bound for power --> upper bound for alpha 
post.odds = 0.95/(1-0.95) 
alpha = prior.odds*pwr/post.odds 
paste(signif(alpha,3)) 
-log10(alpha)
```

```{r}
dd<-dd %>%
   mutate(log10p=-log10(chi_p_Phenotype)) %>% filter(str_detect(Chromosome, "lg"))
chroms<-dd %>% group_by(Chromosome) %>% mutate(Start=min(Index), Stop=max(Index)) %>% 
  select(Chromosome,Start,Stop) %>% unique() %>% 
  mutate(Position=round((Start+Stop)/2))

outliers <- dd %>% filter(log10p >= -log10(alpha))
#Make enough colors
nb.cols <- nrow(chroms)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
#randomize
mycolors<-sample(mycolors)

dd %>% arrange(-log10p)
```

```{r}
ggplot(dd) +
  geom_point(data=dd, aes(x=Index, y=log10p, color=Chromosome), alpha=0.75, cex=0.5) +
  geom_point(data=outliers, aes(x=Index, y=log10p, fill=Chromosome), pch=21, cex=2, alpha=0.9) +
#  geom_hline(yintercept = -log10(0.05/nrow(dd)), col="black", linetype=2, alpha=0.5) +
  geom_hline(yintercept= -log10(alpha), col="black", linetype=1, alpha=0.5) +
  theme_bw() +
  scale_x_continuous(breaks=chroms$Position, labels=chroms$Chromosome) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, face="bold", size=8)) +
  theme(legend.title = element_blank()) +
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  scale_color_manual(values=mycolors) +
  scale_fill_manual(breaks=unique(dd$Chromosome), values=mycolors) +
  ylab("-log10(p)\n") +
  xlab("\nChromosome") +
  ggtitle("Random Assignment Comparison") +
  #labs(subtitle = "Sex as Covariant") +
  theme(plot.title = element_text(hjust=0.5, face="bold") ) +
  theme(plot.subtitle = element_text(hjust=0.5))

```


###Subsample to single classes for dapc

```{r}
fwr<-genind[genind@pop=="FWR"]
fwrx<-as.factor(rep(c("One","Two"),times=c(30,30)))
fwr@pop<-fwrx
fwrdapc<-dapc(fwr, n.pca=55, n.da=1)

scatter(fwrdapc, col=c("purple","green"))
```

### GWAS

Dropping two known problematic samples
```{r}
metaall<-read_csv("metadata/genetics-oto-intersect-06032021_JH.csv") %>%
  filter(!(`Sequence File Name` %in% c("Ht20-30_2012_F04","Ht20-77_2012_E10"))) %>%
  filter(aggregated_new_classes %in% c("FWR","MIG")) 
##  filter(`Birth Year (HOBBS)`==2012)

metaall %>% group_by(aggregated_new_classes) %>% summarize(Count=n())

files<-read_tsv("outputs/100/counts.files.txt", col_names="File")
counts<-read_tsv("outputs/100/counts.txt", col_names="Counts")
counts$Counts<-gsub(" + 0 mapped (100.00% : N/A)", "", counts$Counts, fixed = TRUE)

counted<-bind_cols(files,counts)
counted$`Sequence File Name`<-gsub("_R1.sort.flt.bam","",counted$File)

reads<-left_join(metaall, counted) %>%
  mutate(R1=paste0(`Sequence File Name`,"_R1")) %>% 
  mutate(R2=paste0(`Sequence File Name`,"_R2"))  %>%
  mutate(Pheno=ifelse(aggregated_new_classes=="MIG",0,1))
reads$Counts<-as.numeric(reads$Counts)
```

```{r}
tops<-reads %>% top_frac(.75, Counts) %>% filter(sex!=0)
samplesize<-tops %>% group_by(aggregated_new_classes, `Birth Year (HOBBS)`, sex) %>% summarize(`Sample Size`=n())
samplesize
```

```{r}
ggplot(tops) +
  geom_histogram(aes(Counts)) +
  facet_wrap(.~Pheno) +
  xlim(0, max(tops$Counts))
```

```{r}
bamlist<-tops %>% select(File) %>% mutate(Path=paste0("bams/",File)) %>% select(Path)
write_tsv(bamlist, "bamlists/all-years.bamlist", col_names=FALSE)

phenos<-tops %>% select(Pheno) 
write_tsv(phenos, "phenos/all-years.phenos", col_names=FALSE)

sex<-tops %>% select(sex)
write_tsv(sex, "phenos/all-years.cov", col_names=FALSE)

write_csv(tops, "metadata/all-years-samples.csv")
```

```{sh, eval=FALSE}
srun -p high -t 48:00:00 --mem=32G --nodes=2 $HOME/angsd/angsd -P 24  -bam $HOME/delta-smelt/bamlists/all-years.bamlist \
-yBin $HOME/delta-smelt/phenos/all-years.phenos -cov $HOME/delta-smelt/phenos/all-years.cov -minMapQ 20 -minQ 20 -minInd 131 -doAsso 2 \
-doPost 1 -GL 1 -minCount 2 -out $HOME/delta-smelt/outputs/sanity-check/assoc -doMajorMinor 1 -doMaf 1 -SNP_pval 1e-6 \
-ref $HOME/genomes/hypomesus-20210204/Hyp_tra_F_20210204.fa -rf $HOME/genomes/hypomesus-20210204/large-contigs.txt \
> outputs/sanity-check/assoc.out 2> outputs/sanity-check/assoc.err &
```

```{r}
df<-read_tsv(file="outputs/sanity-check/assoc.lrt0.gz") %>% mutate(log10p = -log10(dchisq(LRT, df = 1))) %>%
  filter(log10p>=0 & log10p != "Inf") %>%
  mutate(p = dchisq(LRT, df=1)) %>%
  mutate(bonfer = p.adjust(p, method = "bonferroni")) %>% 
  mutate(BHfdr = p.adjust(p, method = "BH")) %>%
  group_by(Chromosome) %>%
  mutate(adjpchrom = p.adjust(p, method = "bonferroni")) %>%
  mutate(BHfdrchrom = p.adjust(p, method = "BH"))

df %>% arrange(-LRT) %>% head(n=20)
```


```{r}
variants<-nrow(df)
#Number of expected variants
num<-25
p.T = num/variants 
prior.odds = p.T/(1-p.T) 
pwr = 1 #upper bound for power --> upper bound for alpha 
post.odds = 0.95/(1-0.95) 
alpha = prior.odds*pwr/post.odds 
paste(signif(alpha,3)) 
-log10(alpha)
df %>% dplyr::filter(log10p > -log10(alpha)) 
```

```{r}

data<-df  %>% filter(str_detect(Chromosome, "lg"))
dd <- data %>% ungroup %>% mutate(Index=1:n())
dd$Chromosome<-as.factor(dd$Chromosome)

chroms<-dd %>% group_by(Chromosome) %>% mutate(Start=min(Index), Stop=max(Index)) %>% 
  select(Chromosome,Start,Stop) %>% unique() %>% 
  mutate(Position=round((Start+Stop)/2))

outliers <- dd %>% filter(log10p >= -log10(alpha))
#Make enough colors
nb.cols <- nrow(chroms)
mycolors <- colorRampPalette(brewer.pal(8, "Set1"))(nb.cols)
#randomize
mycolors<-sample(mycolors)

ggplot(dd) +
  geom_point(data=dd, aes(x=Index, y=log10p, color=Chromosome), alpha=0.75, cex=0.5) +
  geom_point(data=outliers, aes(x=Index, y=log10p, fill=Chromosome), pch=21, cex=2, alpha=0.9) +
 # geom_hline(yintercept = -log10(0.05/nrow(dddf)), col="black", linetype=2, alpha=0.5) +
  geom_hline(yintercept= -log10(alpha), col="black", linetype=1, alpha=0.5) +
  theme_bw() +
  scale_x_continuous(breaks=chroms$Position, labels=chroms$Chromosome) +
  theme(axis.text.x = element_text(angle=45, hjust = 1, face="bold", size=8)) +
  theme(legend.title = element_blank()) +
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  scale_color_manual(values=mycolors) +
  scale_fill_manual(breaks=unique(dd$Chromosome), values=mycolors) +
  ylab("-log10(p)\n") +
  xlab("\nChromosome") +
  ggtitle("FWR vs MIG Comparison") +
  labs(subtitle = "Sex as Covariant") +
  theme(plot.title = element_text(hjust=0.5, face="bold") ) +
  theme(plot.subtitle = element_text(hjust=0.5))

ggsave("outputs/sanity-check/all-samples-gwas.jpg", width=8.5, height=11/3)
```

Compare

```{r}
res2021<-read_csv("Supplemental/Supplemental Document S1.csv") %>% select(Chromosome, Position, LRT, log10p) %>% 
  rename(LRT2012=LRT, log10p2012=log10p)
resall<- dd %>% select(Chromosome, Position, LRT, log10p) %>% rename(LRTAll=LRT, log10pAll=log10p)

combo<-left_join(res2021, resall)

top2012<-combo %>% filter(LRT2012 > 15.7) 
topAll<-combo %>% filter(LRTAll > 15.2)

top <- bind_rows(top2012, topAll) %>% unique() %>%
  mutate(Significance2012=ifelse( (LRT2012 > 15.7), "Yes", "No")) %>%
  mutate(SignificanceAll=ifelse( (LRTAll > 15.2), "Yes","No")) %>%
  mutate(SignificanceBoth = ifelse(Significance2012==SignificanceAll, "Yes","No"))

res2021 %>% filter(LRT2012 > 15.7) %>% left_join(resall)
resall %>% filter(LRTAll > 15.2) %>% left_join(res2021)
```

```{r} 
top
```
lg01	2467271	G	A	0.087333	121	16.30721	95/14/0	4.546346	2.842194e-05	
lg02	11230311	A	C	0.305927	121	16.12949	29/22/1	4.505376	3.123377e-05	
lg02	11230464	T	G	0.303122	121	18.22571	62/38/14	4.987094	1.030163e-05	
lg15	2268817	C	T	0.239868	121	15.87289	23/29/0	4.446172	3.579543e-05	
lg23	7996307	G	A	0.147093	121	15.73304	65/22/0	4.413882	3.855835e-05	

```{r}
top
```

```{r}
ggplot(combo, aes(x=log10pAll, y=log10p2012)) +
  geom_point(alpha=.75)+
  geom_vline(xintercept = 4.028243, alpha=0.75, linetype=2) +
  geom_hline(yintercept = 4.006327, alpha=0.75, linetype=2) +
  geom_point(data=top, color="red", alpha=0.75) +
  geom_text_repel(data=top, aes(label=paste0(Chromosome,"-",Position))) +
  theme_bw() +
  theme(panel.grid = element_blank())

ggsave("outputs/sanity-check/gwas-v-gwas.jpg")
```

  